### ISSUE TITLE
Upgrade to PatternFly 6

### ISSUE THREAD

**Issue by ssilvert** [keycloak/keycloak#33710]
### Description

Keycloak login, account, and admin UIs are currently built on top of PatternFly 5. PatternFly 6 was released in October 2024, while PatternFly 5 is still supported it's only a matter of time until PatternFly 7 is released, and at that point PatternFly 5 is no longer supported. The PatternFly 6 update is also required to upgrade from React 18, which is no longer in active support.

As Keycloak 27 will be supported for several years it can not be based on PatternFly 5, and we need to update to PatternFly 6. The downside is this upgrade will break custom themes, and will require some migration effort for anyone upgrading to Keycloak 27.x.

To minimise the impact on the migration Keycloak 26.8.x will be supported for a period of time after Keyckloak 27.0 is released, and we need to provide effort to provide a good migration guide/story around the PatternFly 6 upgrade.

The upgrade will be broken into multiple stages, started after Keycloak 26.8.0 is released:

1. Upgrade Login to PF6 in main
2. Thorough review of all Login pages
3. Upgrade Account to PF6 in main
4. Thorough review of all Account pages
6. Upgrade Admin to PF6 in main
7. Thorough review of all Admin pages
8. Remove dependencies on PF5, including transitive dependencies
9. Provide migration guide/story for custom themes to upgrade to KC 27

### Value Proposition

Upgrade to dependencies that can be supported for the duration of Keycloak 27.x. 

In terms of PF5 vs PF6 and usability/UX we've had mixed responses, where some prefer PF5 while others prefer PF6, and we haven't identified any real value brought by the PF6 upgrade beyond ability to support the UIs.

### Goals

* Update and throughout review of Login UI to PF6
* Update and throughout review of Account UI to PF6
* Update and throughout review of Admin UI to PF6
* Provide a good migration guide/story for migrating custom themes to Keycloak 27
* Reach out to [Keycloakify](https://www.keycloakify.dev/) maintainers, with the aim of having Keycloakify support Keycloak 27 as early as possible

### Notes

At this point we will not consider switching to a different design system such as [Carbon Design System](https://carbondesignsystem.com/) as that will be more effort both for Keycloak maintainers and custom theme developers. This may be considered in the future though.

**Comment by ssilvert** [comment]
Pobably for Keycloak 27

**Comment by bongani-m** [comment]
I've been able to follow the patternfly 6 instructions and created [a fork  with Keycloak to PatternFly v6 in a rough first pass](https://github.com/bongani-m/keycloak/tree/feat/patternfly). There are still some areas that need refinement, but I would say its 90% of the way there. Though I did have a question: this is going to be large PR and I was wondering how to go at this to meet the PR guidelines?

**Screenshots**
![Image](https://github.com/user-attachments/assets/1e408b26-e782-410c-9e16-234880cdd7ce)

![Image](https://github.com/user-attachments/assets/d1f4aaad-f9de-4e18-9970-4a3c4bd900b0)
![Image](https://github.com/user-attachments/assets/a643c0de-9fd9-4d54-a1c7-81dae0160f34)
![Image](https://github.com/user-attachments/assets/aa28a179-658c-42d6-a49e-a42e2d3abaf3)

**Comment by ssilvert** [comment]
> [@ssilvert](https://github.com/ssilvert) I've been able to follow the patternfly 6 upgrade instructions and created [a fork with Keycloak to PatternFly v6 in a rough first pass](https://github.com/bongani-m/keycloak/tree/feat/patternfly). There are still some areas that need refinement, but I would say its 90% of the way there. Though I did have a question: this is going to be large PR and I was wondering how to go at this to meet the PR guidelines?
> 
> **Screenshots** ![Image](https://github.com/user-attachments/assets/1e408b26-e782-410c-9e16-234880cdd7ce)
> 
> ![Image](https://github.com/user-attachments/assets/d1f4aaad-f9de-4e18-9970-4a3c4bd900b0) ![Image](https://github.com/user-attachments/assets/a643c0de-9fd9-4d54-a1c7-81dae0160f34) ![Image](https://github.com/user-attachments/assets/aa28a179-658c-42d6-a49e-a42e2d3abaf3)

@bongani-m Which part(s) of the guidelines are you concerned about?

**Comment by stianst** [comment]
Provisionally setting to 27, but we need to figure out release plans for PatternFly, as well as support for past releases. We also need to figure out the fact that a PatternFly major upgrade is a breaking changes for anyone customising themes. In fact even for those not customising at all it is a big changes as the l&f changes dramatically. That may make it required to introduce PatternFly 6 versions of themes, initially preview then default.

It may end up with something like:

* 26.7 - add experimental/preview PF6 theme versions
* 26.8 - supported PF6 theme versions
* 27.0 - switch to PF6 themes by default, and if migration is acceptable drop support for PF5

**Comment by stianst** [comment]
PatternFly 5 will be supported (important bugs, as well as CVEs) until PatternFly 7 is released, there is no scheduled date for the latter, as the PatternFly team is working on a new minor PF 6 release at the moment.

At the same time we could re-assess if we even want to stick with PatternFly. There are other solutions out there that may be better, for example https://carbondesignsystem.com/.

**Comment by ssilvert** [comment]
> PatternFly 5 will be supported (important bugs, as well as CVEs) until PatternFly 7 is released, there is no scheduled date for the latter, as the PatternFly team is working on a new minor PF 6 release at the moment.
> 
> At the same time we could re-assess if we even want to stick with PatternFly. There are other solutions out there that may be better, for example https://carbondesignsystem.com/.

@stianst @edewit Yes, that's the conclusion we came to after talking to the PatternFly team a few months ago.  We don't have anything forcing us to switch right away so we have time to weigh our options.

**Comment by edewit** [comment]
> At the same time we could re-assess if we even want to stick with PatternFly. There are other solutions out there that may be better, for example https://carbondesignsystem.com/.

Switching a design system would take a long time, it's not a simple replace the components. Even though they have similar components the properties and how they work will not be the same. Also the layout system is complexly different. We would need a new UI design to fit this design system.

**Comment by ssilvert** [comment]
> > At the same time we could re-assess if we even want to stick with PatternFly. There are other solutions out there that may be better, for example https://carbondesignsystem.com/.
> 
> Switching a design system would take a long time, it's not a simple replace the components. Even though they have similar components the properties and how they work will not be the same. Also the layout system is complexly different. We would need a new UI design to fit this design system.

Yeah, I agree 100%.

If PatternFly is a dead end then we will have to switch eventually.  But "eventually" won't come for quite some time.
